### ISSUE TITLE
ClickHouse sometimes fails to parse `DateTime64` columns using `JSONEachRow` depending on the position of keys within the JSON input

### ISSUE THREAD

**Issue by michaelsmithxyz** [ClickHouse/ClickHouse#36949]
ClickHouse sometimes fails to parse `DateTime64` columns using `JSONEachRow` depending on the position of keys within the JSON input.

**How to reproduce**

Using ClickHouse 22.4.4 over the HTTP interface, with the following table and `date_time_input_format=basic` (the default):
```
CREATE TABLE test (
   `i` Int64,
  `d` DateTime64
) ENGINE Log;
```

You'll see inconsistent behavior when attempting to insert values into `d` with the `JSONEachRow` input format. These work:
```
INSERT INTO test FORMAT JSONEachRow
{"d": "2022-05-03", "i": 123};
```
```
INSERT INTO test FORMAT JSONEachRow
{"i": 123, "d": "2022-05-03 00:00:00"};
```

However, this does not:
```
INSERT INTO test FORMAT JSONEachRow
{"i": 123, "d": "2022-05-03"};
```
```
Code: 27. DB::ParsingException: Cannot parse input: expected '"' at end of stream.: (while reading the value of key d): While executing ParallelParsingBlockInputFormat: (at row 1)
. (CANNOT_PARSE_INPUT_ASSERTION_FAILED) (version 22.4.4.7 (official build))
```

In other words, it seems like ClickHouse can't parse `DateTime64` columns from JSON in the `YYYY-MM-DD` format if they appear in the last position of the incoming object. I'd expect all of these queries to work.

**Comment by filimonov** [comment]
The same is true for the plain DateTime.

**Comment by michaelsmithxyz** [comment]
This also doesn't seem strictly limited to `JSONEachRow`. I haven't tested every JSON input format but, for example, under the same conditions as above:

This works:
```
insert into abc format JSONCompactEachRowWithNames
["d"]["2017-08-31 00:00:00"];
```
This doesn't (different error message though):
```
insert into abc format JSONCompactEachRowWithNames
["d"]["2017-08-31"];
```
```
Code: 27. DB::ParsingException: Cannot parse input: expected '"' at end of stream.: (while reading the value of key d): Buffer has gone, cannot extract information about what has been parsed.: While executing ParallelParsingBlockInputFormat: (at row 1)
. (CANNOT_PARSE_INPUT_ASSERTION_FAILED) (version 22.4.4.7 (official build))
```
That example doesn't seem to be order-dependent though since this also fails with the error:
```
insert into abc format JSONCompactEachRowWithNames
["d", "i"]["2017-08-31", 1];
```

**Comment by alexey-milovidov** [comment]
I confirm. This issue is very easy to reproduce:
```
milovidov-desktop :) CREATE TEMPORARY TABLE test (
                        `i` Int64,
                       `d` DateTime64
                     )

CREATE TEMPORARY TABLE test
(
    `i` Int64,
    `d` DateTime64
)

Query id: 2b98059a-36df-4e79-a65f-f28fefa59e48

Ok.

0 rows in set. Elapsed: 0.000 sec. 

milovidov-desktop :) INSERT INTO test FORMAT JSONEachRow
                     {"i": 123, "d": "2022-05-03"};

INSERT INTO test FORMAT JSONEachRow

Query id: 98c22b4f-b4c9-42e8-8ecf-f69688000af1

Ok.
Exception on client:
Code: 27. DB::ParsingException: Cannot parse input: expected '"' at end of stream.: (while reading the value of key d): While executing ParallelParsingBlockInputFormat: data for INSERT was parsed from query: (at row 1)
. (CANNOT_PARSE_INPUT_ASSERTION_FAILED)
```

Although it is not a bug, because `2022-05-03` is just Date, not DateTime64.
Let's check why this is happening and how to improve.

PS. Unassigned @Enmk because nothing is done.
